src/java.base/share/classes/sun/security/ssl/SessionTicketExtension.java

```diff
@@ -145,79 +164,22 @@ private static final class KeyState {
 
         // Get a key with a specific key number
         static StatelessKey getKey(HandshakeContext hc, int num)  {
-            StatelessKey ssk = hc.sslContext.keyHashMap.get(num);
+            StatelessKey ssk = hc.sslContext.getKey(num);
 
-            if (ssk == null || ssk.isInvalid(getSessionTimeout(hc))) {
+            if (ssk == null || ssk.isInvalid(hc.sslContext)) {
                 return null;
             }
             return ssk;
         }
 
         // Get the current valid key, this will generate a new key if needed
         static StatelessKey getCurrentKey(HandshakeContext hc) {
-            StatelessKey ssk = hc.sslContext.keyHashMap.get(currentKeyID);
+            StatelessKey ssk = hc.sslContext.getKey();
 
             if (ssk != null && !ssk.isExpired()) {
                 return ssk;
             }
-            return nextKey(hc);
-        }
-
-        // This method locks when the first getCurrentKey() finds it to be too
-        // old and create a new key to replace the current key.  After the new
-        // key established, the lock can be released so following
-        // operations will start using the new key.
-        // The first operation will take a longer code path by generating the
-        // next key and cleaning up old keys.
-        private static StatelessKey nextKey(HandshakeContext hc) {
-            StatelessKey ssk;
-
-            synchronized (hc.sslContext.keyHashMap) {
-                // If the current key is no longer expired, it was already
-                // updated by a previous operation, and we can return.
-                ssk = hc.sslContext.keyHashMap.get(currentKeyID);
-                if (ssk != null && !ssk.isExpired()) {
-                    return ssk;
-                }
-                int newNum;
-                if (currentKeyID == Integer.MAX_VALUE) {
-                    newNum = 0;
-                } else {
-                    newNum = currentKeyID + 1;
-                }
```

Comment:
Any idea why the code used zero after MAX_VALUE before?

Reply:
I have no clue, but as the keys can be initially negative anyway (i.e. if `SecureRandom().nextInt()` returns a negative number), I don't see a reason to restrict them to positive values after they wrap around.